calico_node/startup/startup.go

```diff
@@ -128,6 +130,31 @@ func main() {
 	message("Using node name: %s", nodeName)
 }
 
+// configureNodeRef will attempt to discover the cluster type it is running on, check to ensure we
+// have not already set it on this Node, and set it if need be.
+func configureNodeRef(node *api.Node) {
+	orchestrator := ""
+	nodeRef := ""
+	// Sort out what type of cluster we're running on.
+	if nodeRef = os.Getenv("CALICO_K8S_NODE_REF"); nodeRef != "" {
+		orchestrator = "k8s"
+	} else {
+		return
+	}
+
+	orchRef := api.OrchRef{
+		NodeName:     nodeRef,
+		Orchestrator: orchestrator,
+	}
+	// Return if we find that this OrchRef is not already set on the Node.
+	for _, ref := range node.Spec.OrchRefs {
```

Comment:
Probably fine, but given we know we only support this on k8s we could do a simple `if len(node.Spec.OrchRefs) != 0`.

Don't care too much either way.

Reply:
A simpler approach might just be to write this directly every time.

This is slightly different, in that we'll actually update the orch ref if it changes. Unclear if that is more or less desirable. I think it is probably better.

```go
node.Spec.OrchRefs = []api.OrchRef{api.OrchRef{. . .}}
```

Reply:
I think I agree, it would make it a bit more straightforward to expand on later, but I don't know if there are any repercussions if we inadvertently remove some OrchRefs, but I don't _think_ that is a problem.